### PR TITLE
Integrate MyCLI as MySQL Completer

### PR DESCRIPTION
**Pull request by KevinMicrosoft**
Integrated [MyCLI](https://github.com/dbcli/mycli/tree/master/mycli)

* Copied mycli completer here as MySQLCompleter, but removed CLI-specific features, such as special commands, favorite query, table format, and filenames completions. Copied over all naive and smart completion tests.

* Placed type suggestion functions in new file mysql_completion_engine.

* Created new mysqlliterals.json, taken from MySQL docs and connected to new MySQLCompleter. I also added well-known follow up keyword handling to completer.

* Created new mysql_utils folder within parseutils and copied over parse utils. Any functions that were the same, i placed in parseutils/utils.py. Copied over parseutils tests

* Copied mycli completion_refresher functions into completion_refresher.py. Created a new refreshers map, so service will grab correct list of refresher functions based on provider name.

* Copied over mycli query executor into mysql_lightweight_metadata

### CODE REVIEW

File: ossdbtoolsservice/language/query/mysql_lightweight_metadata.py [hunk 1]
```diff
@@ -4,78 +4,102 @@
...
+
+    version_comment_query = '''SELECT @@VERSION_COMMENT'''
+    version_comment_query_mysql4 = '''SHOW VARIABLES LIKE "version_comment"'''
+
+    show_candidates_query = '''SELECT name from mysql.help_topic WHERE name like "SHOW %"'''
+
+    users_query = '''SELECT CONCAT("'", user, "'@'",host,"'") FROM mysql.user'''
+
+    functions_query = '''SELECT ROUTINE_NAME FROM INFORMATION_SCHEMA.ROUTINES
+    WHERE ROUTINE_TYPE="FUNCTION" AND ROUTINE_SCHEMA = "%s"'''
```
Comment (swjain23) [inline ossdbtoolsservice/language/query/mysql_lightweight_metadata.py:28]: fix the indentation

File: tests/language/completion/mysql_completion/test_smart_completion_public_schema_only.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+                Document(text=text, cursor_position=position),
+                self.complete_event))
+        self.assertSetEqual(result, set(map(lambda completion: Completion(completion, display_meta='keyword'), sorted(self.keywords))))
+
+
+    def test_select_keyword_completion(self):
+        text = 'SEL'
+        position = len('SEL')
+        result = set(self.completer.get_completions(
+            Document(text=text, cursor_position=position),
```
Comment (swjain23) [inline tests/language/completion/mysql_completion/test_smart_completion_public_schema_only.py:51]: Can you add comments to these tests mentioning what is exactly happening. I have seen it at other places, that makes it easier to understand.

File: tests/language/completion/pg_completion/test_naive_completion.py [hunk 3]
```diff
@@ -80,38 +80,3 @@ def test_alter_well_known_keywords_completion(self):
             Completion(text="SYSTEM", display_meta='keyword'),
         ]))
         self.assertTrue(Completion(text="CREATE", display_meta="keyword") not in result)
-
-    def test_keyword_lower_casing(self):
```
Comment (swjain23) [inline tests/language/completion/pg_completion/test_naive_completion.py:84]: Why are we removing these tests?

Comment (KevinMicrosoft) [reply tests/language/completion/pg_completion/test_naive_completion.py:84]: Casing is only checked if smart completion is on, so I moved these tests so the smart completion test file.

File: tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py [hunk 4]
```diff
@@ -865,38 +865,40 @@ def test_set_schema(self, completer):
         self.assertEqual(result, expected)
 
     def test_keyword_lower_casing(self, ):
-        new_completer = PGCompleter(smart_completion=True, settings={'keyword_casing':'lower'})
-
+        new_completer = PGCompleter(smart_completion=True, settings={'keyword_casing': 'lower'})
         text = 'SEL'
         position = len(text)
+
+        # When I request completions from SEL with keyword_casing as 'upper'
```
Comment (swjain23) [inline tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py:872]: It is lower

File: tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py [hunk 5]
```diff
@@ -865,38 +865,40 @@ def test_set_schema(self, completer):
...
         result = set(new_completer.get_completions(
             Document(text=text, cursor_position=position),
             Mock()))
 
-        # then completions should now be lower case
+        # Then completions should now be upper case
```
Comment (swjain23) [inline tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py:877]: It should be lower as you are matching with lower case select

File: tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py [hunk 6]
```diff
@@ -865,38 +865,40 @@ def test_set_schema(self, completer):
...
         self.assertSetEqual(result, set([Completion(text='select', start_position=-3, display_meta="keyword")]))
 
     def test_keyword_upper_casing(self, ):
-        new_completer = PGCompleter(smart_completion=True, settings={'keyword_casing':'upper'})
-
+        new_completer = PGCompleter(smart_completion=True, settings={'keyword_casing': 'upper'})
         text = 'sel'
         position = len(text)
+
+        # When I request completions from sel with keyword_casing as 'lower'
```
Comment (swjain23) [inline tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py:885]: This is upper

File: tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py [hunk 7]
```diff
@@ -865,38 +865,40 @@ def test_set_schema(self, completer):
...
         result = set(new_completer.get_completions(
             Document(text=text, cursor_position=position),
             Mock()))
 
-        # then completions should now be lower case
+        # Then completions should now be lower case
```
Comment (swjain23) [inline tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py:890]: This is upper

File: tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py [hunk 8]
```diff
@@ -865,38 +865,40 @@ def test_set_schema(self, completer):
...
-    def test_keyword_auto_casing(self, ):        
-        new_completer = PGCompleter(smart_completion=True, settings={'keyword_casing':'auto'})
-
-        # if text is lower case   
+    def test_keyword_auto_casing(self, ):
+        new_completer = PGCompleter(smart_completion=True, settings={'keyword_casing': 'auto'})
         text = 'sel'
         position = len(text)
+
+        # When I request completions from sel with keyword_casing as 'auto'
```
Comment (swjain23) [inline tests/language/completion/pg_completion/test_smart_completion_public_schema_only.py:898]: mention that as text is lower, the completion is lower as well. As it depends on actual text when it is auto

**Pull request closed without merging**